Faces 4.0: JSF 2.3 FAT tests fail on Java 17+ when running with the EE10 repeat action

**Latest exception that remains is: https://github.com/OpenLiberty/open-liberty/issues/22631#issuecomment-1261234580 going to disable these specific tests for now and come back to them.**


Plan is to disable the EE10 bucket execution on Java 17+ until we have a solution.

An example of some of the failures we need to debug:

1)` Caused by: java.lang.ClassCastException: org.apache.myfaces.core.api.shared.lang.PropertyDescriptorWrapper incompatible with org.apache.myfaces.core.api.shared.lang.LambdaPropertyDescriptor`
2) `java.lang.invoke.LambdaConversionException: Invalid caller:`

The first exception does pop up on Java11 in a test that we have not enabled for EE10 yet. The second issue seems to be due to a new Java14+ check:

> 
>        if (!caller.hasFullPrivilegeAccess()) {
>             throw new LambdaConversionException(String.format(
>                     "Invalid caller: %s",
>                     caller.lookupClass().getName()));
>         }

>         /**
>          * Returns {@code true} if this lookup has <em>full privilege access</em>,
>          * i.e. {@code PRIVATE} and {@code MODULE} access.
>          * A {@code Lookup} object must have full privilege access in order to
>          * access all members that are allowed to the
>          * {@linkplain #lookupClass() lookup class}.
>          *
>          * @return {@code true} if this lookup has full privilege access.
>          * @since 14
>          * @see <a href="MethodHandles.Lookup.html#privacc">private and module access</a>
>          */
>         public boolean hasFullPrivilegeAccess() {
>             return (allowedModes & (PRIVATE|MODULE)) == (PRIVATE|MODULE);
>         }

MyFaces code in question (privateLookupIn, potentially isn't doing what was expected)

>          MethodHandles.Lookup lookup = (MethodHandles.Lookup)
>          privateLookupIn.invoke(null, target, MethodHandles.lookup());
>         for (PropertyDescriptor pd : Introspector.getBeanInfo(target).getPropertyDescriptors())
>         {
>             LambdaPropertyDescriptor lpd = createLambdaPropertyDescriptor(target, pd, lookup);

The second exception is just logged it looks and we fall back to the old way of doing things before Faces 4.0. However, we should still look at it so we can improve this functionality. The first issue causes an FFDC to happen so I think that one needs investigating first.

Current failure:

> 
> Stack Dump = jakarta.servlet.ServletException: argument type mismatch
> 	at jakarta.faces.webapp.FacesServlet.service(FacesServlet.java:255)
> 	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1258)
> 	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:746)
> 	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:443)
> 	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1227)
> 	at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:5058)
> 	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.handleRequest(DynamicVirtualHost.java:316)
> 	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:1007)
> 	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:281)
> 	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1239)
> 	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:468)
> 	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:427)
> 	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:567)
> 	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:501)
> 	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:361)
> 	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:328)
> 	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:167)
> 	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:75)
> 	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:514)
> 	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:584)
> 	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:968)
> 	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1057)
> 	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:245)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
> 	at java.base/java.lang.Thread.run(Thread.java:858)
> Caused by: java.lang.IllegalArgumentException: argument type mismatch
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
> 	at jakarta.faces.component._ComponentAttributesMap.setComponentProperty(_ComponentAttributesMap.java:688)
> 	at jakarta.faces.component._ComponentAttributesMap.put(_ComponentAttributesMap.java:582)
> 	at jakarta.faces.component._ComponentAttributesMap.put(_ComponentAttributesMap.java:59)
> 	at org.apache.myfaces.view.facelets.tag.faces.ComponentRule$LiteralAttributeMetadata.applyMetadata(ComponentRule.java:53)
> 	at org.apache.myfaces.view.facelets.tag.MetadataImpl.applyMetadata(MetadataImpl.java:45)
> 	at jakarta.faces.view.facelets.MetaTagHandler.setAttributes(MetaTagHandler.java:63)
> 	at jakarta.faces.view.facelets.DelegatingMetaTagHandler.setAttributes(DelegatingMetaTagHandler.java:93)
> 	at org.apache.myfaces.view.facelets.tag.faces.ComponentTagHandlerDelegate.apply(ComponentTagHandlerDelegate.java:281)
> 	at jakarta.faces.view.facelets.DelegatingMetaTagHandler.apply(DelegatingMetaTagHandler.java:52)
> 	at jakarta.faces.view.facelets.CompositeFaceletHandler.apply(CompositeFaceletHandler.java:47)
> 	at jakarta.faces.view.facelets.DelegatingMetaTagHandler.applyNextHandler(DelegatingMetaTagHandler.java:57)
> 	at org.apache.myfaces.view.facelets.tag.faces.ComponentTagHandlerDelegate.apply(ComponentTagHandlerDelegate.java:362)
> 	at jakarta.faces.view.facelets.DelegatingMetaTagHandler.apply(DelegatingMetaTagHandler.java:52)
> 	at jakarta.faces.view.facelets.DelegatingMetaTagHandler.applyNextHandler(DelegatingMetaTagHandler.java:57)
> 	at org.apache.myfaces.view.facelets.tag.faces.ComponentTagHandlerDelegate.apply(ComponentTagHandlerDelegate.java:362)
> 	at jakarta.faces.view.facelets.DelegatingMetaTagHandler.apply(DelegatingMetaTagHandler.java:52)
> 	at jakarta.faces.view.facelets.CompositeFaceletHandler.apply(CompositeFaceletHandler.java:47)
> 	at org.apache.myfaces.view.facelets.compiler.NamespaceHandler.apply(NamespaceHandler.java:67)
> 	at jakarta.faces.view.facelets.CompositeFaceletHandler.apply(CompositeFaceletHandler.java:47)
> 	at jakarta.faces.view.facelets.CompositeFaceletHandler.apply(CompositeFaceletHandler.java:47)
> 	at org.apache.myfaces.view.facelets.compiler.EncodingHandler.apply(EncodingHandler.java:49)
> 	at org.apache.myfaces.view.facelets.impl.DefaultFacelet.apply(DefaultFacelet.java:195)
> 	at org.apache.myfaces.view.facelets.FaceletViewDeclarationLanguage.buildView(FaceletViewDeclarationLanguage.java:423)
> 	at org.apache.myfaces.lifecycle.RenderResponseExecutor.execute(RenderResponseExecutor.java:83)
> 	at org.apache.myfaces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:241)
> 	at jakarta.faces.webapp.FacesServlet.service(FacesServlet.java:225)
> 	... 25 more

--- a/dev/com.ibm.ws.jsf.2.3_fat/fat/src/com/ibm/ws/jsf23/fat/tests/JSF23FacesDataModelTests.java
+++ b/dev/com.ibm.ws.jsf.2.3_fat/fat/src/com/ibm/ws/jsf23/fat/tests/JSF23FacesDataModelTests.java
@@ -28,6 +28,7 @@ import com.ibm.websphere.simplicity.log.Log;
 import com.ibm.ws.jsf23.fat.JSFUtils;
 
 import componenttest.annotation.Server;
+import componenttest.annotation.SkipForRepeat;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.topology.impl.LibertyServer;
 
@@ -71,6 +72,7 @@ public class JSF23FacesDataModelTests {
      * @throws Exception
      */
     @Test
+    @SkipForRepeat(SkipForRepeat.EE10_FEATURES) // Test fails on Java17+, needs more investigation.
     public void testFacesDataModelUIRepeat() throws Exception {
         String contextRoot = "FacesDataModel";
         try (WebClient webClient = new WebClient()) {
@@ -191,6 +193,7 @@ public class JSF23FacesDataModelTests {
      * @throws Exception
      */
     @Test
+    @SkipForRepeat(SkipForRepeat.EE10_FEATURES) // Test fails on Java17+, needs more investigation.
     public void testFacesDataModelChildUIRepeat() throws Exception {
         String contextRoot = "FacesDataModel";
         try (WebClient webClient = new WebClient()) {

--- a/dev/com.ibm.ws.jsf.2.3_fat/fat/src/com/ibm/ws/jsf23/fat/tests/JSF23SelectOneRadioGroupTests.java
+++ b/dev/com.ibm.ws.jsf.2.3_fat/fat/src/com/ibm/ws/jsf23/fat/tests/JSF23SelectOneRadioGroupTests.java
@@ -702,6 +702,7 @@ public class JSF23SelectOneRadioGroupTests {
      * @throws Exception
      */
     @Test
+    @SkipForRepeat(SkipForRepeat.EE10_FEATURES)
     public void testSelectOneRadioGroup_uiRepeat() throws Exception {
         try (WebClient webClient = new WebClient()) {
 

--- a/dev/com.ibm.ws.jsf.2.3_fat/fat/src/com/ibm/ws/jsf23/fat/tests/JSF23UIRepeatConditionTests.java
+++ b/dev/com.ibm.ws.jsf.2.3_fat/fat/src/com/ibm/ws/jsf23/fat/tests/JSF23UIRepeatConditionTests.java
@@ -148,6 +148,7 @@ public class JSF23UIRepeatConditionTests {
     @Mode(TestMode.FULL)
     @Test
     @ExpectedFFDC({ "javax.servlet.ServletException" })
+    @SkipForRepeat(SkipForRepeat.EE10_FEATURES)
     public void testUIRepeatConditionErrorEndTooLarge() throws Exception {
         String contextRoot = "UIRepeatConditionCheck";
         String errorText = "end cannot be greater than collection size";

